File: src/main/java/com/gtladd/gtladditions/api/registry/GTLAddMultiBlockMachineBuilder.java
package com.gtladd.gtladditions.api.registry;

import org.gtlcore.gtlcore.common.data.GTLMachines;
import org.gtlcore.gtlcore.utils.TextUtil;

import com.gregtechceu.gtceu.api.block.IMachineBlock;
import com.gregtechceu.gtceu.api.data.RotationState;
import com.gregtechceu.gtceu.api.item.MetaMachineItem;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.MultiblockMachineDefinition;
import com.gregtechceu.gtceu.api.machine.multiblock.MultiblockControllerMachine;
import com.gregtechceu.gtceu.api.registry.registrate.MultiblockMachineBuilder;

import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;

import org.apache.commons.lang3.function.TriFunction;

import java.util.function.BiFunction;
import java.util.function.Function;

public class GTLAddMultiBlockMachineBuilder extends MultiblockMachineBuilder {

    private GTLAddMultiBlockMachineBuilder(String name, Function<IMachineBlockEntity, ? extends MultiblockControllerMachine> metaMachine,
                                           BiFunction<BlockBehaviour.Properties, MultiblockMachineDefinition, IMachineBlock> blockFactory,
                                           BiFunction<IMachineBlock, Item.Properties, MetaMachineItem> itemFactory,
                                           TriFunction<BlockEntityType<?>, BlockPos, BlockState, IMachineBlockEntity> blockEntityFactory) {
        super(GTLAddRegistration.REGISTRATE, name, metaMachine, blockFactory, itemFactory, blockEntityFactory);
    }

    public static GTLAddMultiBlockMachineBuilder createMulti(String name, Function<IMachineBlockEntity, ? extends MultiblockControllerMachine> metaMachine,
                                                             BiFunction<BlockBehaviour.Properties, MultiblockMachineDefinition, IMachineBlock> blockFactory,
                                                             BiFunction<IMachineBlock, Item.Properties, MetaMachineItem> itemFactory,
                                                             TriFunction<BlockEntityType<?>, BlockPos, BlockState, IMachineBlockEntity> blockEntityFactory) {
        return new GTLAddMultiBlockMachineBuilder(name, metaMachine, blockFactory, itemFactory, blockEntityFactory);
    }

    public GTLAddMultiBlockMachineBuilder allRotation() {
        return (GTLAddMultiBlockMachineBuilder) super.rotationState(RotationState.ALL);
    }

    public GTLAddMultiBlockMachineBuilder nonYAxisRotation() {
        return (GTLAddMultiBlockMachineBuilder) super.rotationState(RotationState.NON_Y_AXIS).allowExtendedFacing(false);
    }

    public GTLAddMultiBlockMachineBuilder noneRotation() {
        return (GTLAddMultiBlockMachineBuilder) super.rotationState(RotationState.NONE).allowExtendedFacing(false).allowFlip(false);
    }

    public GTLAddMultiBlockMachineBuilder tooltipText(String string) {
        return (GTLAddMultiBlockMachineBuilder) super.tooltips(new Component[] { Component.literal(string) });
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextCoilParallel() {
        return tooltipText("线圈温度每增加900K，并行数X2");
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextLaser() {
        return tooltipText("允许使用激光仓");
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextMultiRecipes() {
        return tooltipText("支持跨配方并行");
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextSuperMultiRecipes() {
        return tooltipText("拥有特殊的跨配方并行");
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextParallelHatch() {
        return tooltipText("通过并行控制仓让机器同时处理多个相同配方");
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextPerfectOverclock() {
        return tooltipText("超频不会损失能耗");
    }

    public GTLAddMultiBlockMachineBuilder tooltipTextAdd() {
        return (GTLAddMultiBlockMachineBuilder) super.tooltips(Component.literal(TextUtil.full_color("由GTLAdditions添加")));
    }

    public GTLAddMultiBlockMachineBuilder coilparalleldisplay() {
        return (GTLAddMultiBlockMachineBuilder) super.additionalDisplay(GTLMachines.MULTIPLERECIPES_COIL_PARALLEL);
    }
}

File: src/main/java/com/gtladd/gtladditions/api/registry/GTLAddMultiBlockMachineBuilder.kt
package com.gtladd.gtladditions.api.registry

import com.gregtechceu.gtceu.api.block.IMachineBlock
import com.gregtechceu.gtceu.api.data.RotationState
import com.gregtechceu.gtceu.api.item.MetaMachineItem
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.MultiblockMachineDefinition
import com.gregtechceu.gtceu.api.machine.multiblock.MultiblockControllerMachine
import com.gregtechceu.gtceu.api.recipe.GTRecipeType
import com.gregtechceu.gtceu.api.registry.registrate.MultiblockMachineBuilder
import com.gtladd.gtladditions.utils.ComponentExtensions.toComponent
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockBehaviour
import net.minecraft.world.level.block.state.BlockState
import org.apache.commons.lang3.function.TriFunction
import org.gtlcore.gtlcore.common.data.GTLMachines
import java.util.function.BiFunction
import java.util.function.Function

class GTLAddMultiBlockMachineBuilder private constructor(
    name: String,
    metaMachine: Function<IMachineBlockEntity, out MultiblockControllerMachine>,
    blockFactory: BiFunction<BlockBehaviour.Properties, MultiblockMachineDefinition, IMachineBlock>,
    itemFactory: BiFunction<IMachineBlock, Item.Properties, MetaMachineItem>,
    blockEntityFactory: TriFunction<BlockEntityType<*>, BlockPos, BlockState, IMachineBlockEntity>
) : MultiblockMachineBuilder(
    GTLAddRegistration.REGISTRATE,
    name,
    metaMachine,
    blockFactory,
    itemFactory,
    blockEntityFactory
) {

    fun allRotation(): GTLAddMultiBlockMachineBuilder = super.rotationState(RotationState.ALL) as GTLAddMultiBlockMachineBuilder

    fun nonYAxisRotation(): GTLAddMultiBlockMachineBuilder = super.rotationState(RotationState.NON_Y_AXIS).allowExtendedFacing(false) as GTLAddMultiBlockMachineBuilder

    fun noneRotation(): GTLAddMultiBlockMachineBuilder = super.rotationState(RotationState.NONE).allowExtendedFacing(false).allowFlip(false) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextKey(vararg key: Component): GTLAddMultiBlockMachineBuilder = super.tooltips(*key) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextMaxParallels(parallel: Any): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.multiblock.max_parallel".toComponent(parallel)) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextRecipeTypes(vararg recipeTypes: GTRecipeType): GTLAddMultiBlockMachineBuilder {
        val components = recipeTypes.map { it.registryName.toLanguageKey().toComponent }.toTypedArray()
        return super.tooltips("gtceu.machine.available_recipe_map_${recipeTypes.size}.tooltip".toComponent(*components)) as GTLAddMultiBlockMachineBuilder
    }

    fun tooltipTextCoilParallel(): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.multiblock.coil_parallel".toComponent) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextLaser(): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.multiblock.laser.tooltip".toComponent) as GTLAddMultiBlockMachineBuilder

    fun tooltipOnlyTextLaser(): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.multiblock.only.laser.tooltip".toComponent) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextMultiRecipes(): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.machine.multiple_recipes.tooltip".toComponent) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextParallelHatch(): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.multiblock.parallelizable.tooltip".toComponent) as GTLAddMultiBlockMachineBuilder

    fun tooltipTextPerfectOverclock(): GTLAddMultiBlockMachineBuilder = super.tooltips("gtceu.machine.perfect_oc".toComponent) as GTLAddMultiBlockMachineBuilder

    fun coilParallelDisplay(): GTLAddMultiBlockMachineBuilder = super.additionalDisplay(GTLMachines.MULTIPLERECIPES_COIL_PARALLEL) as GTLAddMultiBlockMachineBuilder

    companion object {
        fun createMulti(
            name: String,
            metaMachine: Function<IMachineBlockEntity, out MultiblockControllerMachine>,
            blockFactory: BiFunction<BlockBehaviour.Properties, MultiblockMachineDefinition, IMachineBlock>,
            itemFactory: BiFunction<IMachineBlock, Item.Properties, MetaMachineItem>,
            blockEntityFactory: TriFunction<BlockEntityType<*>, BlockPos, BlockState, IMachineBlockEntity>
        ): GTLAddMultiBlockMachineBuilder = GTLAddMultiBlockMachineBuilder(name, metaMachine, blockFactory, itemFactory, blockEntityFactory)
    }
}